### PR TITLE
Move to Se 3.0.1 release.

### PR DESCRIPTION
**Pull request by mach6**
- [X] By placing an `X` in the preceding checkbox, I verify that I am a PayPal employee or
  I have signed the [Contributor License Agreement](https://github.com/paypal/SeLion#contributing)

---

Note: I left this in 3 commits as an attempt to group the changes.
- Upgrade to the latest Se3 release
- Update how launcher configurations are
  passed / initialized
- Update commons-io and htmlunit versions
- Update and Add tests
- default to marionette true
- pay attention to webdriver.firefox.marionette
- provide path to boot local node with legacy ff support
- delete GeckoCapabilitiesBuilder --> move functionality
  to FirefoxCapabilitiesBuilder
- rename TestCapabilitiyBuilder which is specific to Travis
- update ci scripts

### CODE REVIEW

File: client/src/test/java/com/paypal/selion/TravisCICapabilityBuilder.java [hunk 1]
```diff
@@ -12,39 +12,39 @@
 |  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for  |
```
Comment (mach6) [inline client/src/test/java/com/paypal/selion/TravisCICapabilityBuilder.java:12]: Note: `GeckoCapabilitiesBuilder` was not renamed. it was deleted... Git or github issue here ... The diff for this file is not right.. `TestCapabilityBuilder` was renamed to `TravisCICapabilityBuilder`

File: server/src/main/java/com/paypal/selion/grid/AbstractBaseLauncher.java [hunk 2]
```diff
@@ -60,6 +84,12 @@
      */
     private InstanceType type;
 
+    void ident() {
+        System.out.print(SEPARATOR);
+        System.out.print(SeLionBuildInfo.getBuildValue(SeLionBuildInfo.SeLionBuildProperty.SELION_VERSION));
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/AbstractBaseLauncher.java]: will not fix

File: server/src/main/java/com/paypal/selion/grid/AbstractBaseLauncher.java [hunk 3]
```diff
@@ -60,6 +84,12 @@
      */
     private InstanceType type;
 
+    void ident() {
+        System.out.print(SEPARATOR);
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/AbstractBaseLauncher.java]: will not fix

File: server/src/main/java/com/paypal/selion/grid/SelendroidJarSpawner.java [hunk 4]
```diff
@@ -93,4 +94,31 @@ private CommandLine createJavaCommandForChildProcess() throws IOException {
...
+        usage.append("  Driver Options: \n");
+        usage.append("    Any valid Selendroid dash option(s). \n");
+        usage.append("\n");
+        usage.append("  System Properties: \n");
+        usage.append("    -DselionHome=<folderPath>: \n");
+        usage.append("       Path of SeLion home directory. Defaults to <user.home>/.selion/ \n");
+        usage.append("    -D[property]=[value]: \n");
+        usage.append("       Any other System Property you wish to pass to the JVM \n");
+
+        System.out.print(usage.toString());
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/SelendroidJarSpawner.java]: will not fix

File: server/src/test/java/com/paypal/selion/grid/AbstractBaseLauncherTest.java [hunk 5]
```diff
@@ -123,23 +130,25 @@ public void testGetProgramArguments() throws Exception {
         assertTrue(Arrays.asList(launcher.getProgramArguments()).contains(SELION_CONFIG_FILE));
 
         // "node" case
-        launcher = new DummyProcessLauncher(new String[] {ROLE_ARG, "node"});;
+        launcher = new DummyProcessLauncher(new String[] { ROLE_ARG, "node" });
+        ;
```
Comment (mach6) [inline server/src/test/java/com/paypal/selion/grid/AbstractBaseLauncherTest.java]: WIll fix this.

File: server/src/main/java/com/paypal/selion/grid/AppiumSpawner.java [hunk 6]
```diff
@@ -75,4 +76,33 @@ private CommandLine createCommandForChildProcess() throws IOException {
...
+        usage.append("  Driver Options: \n");
+        usage.append("    Any valid Appium dash option(s). \n");
+        usage.append("\n");
+        usage.append("  System Properties: \n");
+        usage.append("    -DselionHome=<folderPath>: \n");
+        usage.append("       Path of SeLion home directory. Defaults to <user.home>/.selion/ \n");
+        usage.append("    -D[property]=[value]: \n");
+        usage.append("       Any other System Property you wish to pass to the JVM \n");
+
+        System.out.print(usage.toString());
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/AppiumSpawner.java]: will not fix

**Comment by mach6** [comment]
@sebady please review

File: server/pom.xml [hunk 7]
```diff
@@ -153,6 +153,12 @@
         </dependency>
 
         <dependency>
+            <groupId>com.beust</groupId>
+            <artifactId>jcommander</artifactId>
+            <version>1.48</version>
```
Comment (sebady) [inline server/pom.xml]: Can we manage this version in Project BOM?

File: server/src/main/java/com/paypal/selion/grid/JarSpawner.java [hunk 8]
```diff
@@ -52,20 +51,9 @@ public static void main(String[] args) {
...
-        usage.append("\n");
-        usage.append("  Options:\n");
-        usage.append("    Any valid standalone or grid dash options. See above.\n");
-        usage.append("       E.g. '-role hub -port 2054' -- To start a hub on port 2054. \n");
-        usage.append("\n");
         usage.append("  System Properties: \n");
         usage.append("    -DselionHome=<folderPath>: \n");
-        usage.append("       Path of SeLion home directory. Defaults to \n");
-        usage.append("       <user.home>/.selion/ \n");
+        usage.append("       Path of SeLion home directory. Defaults to <user.home>/.selion/ \n");
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/JarSpawner.java]: Should we consider defaulting to "<user.home>/.selion2" For SeLion 2 ?  To allow SeLIon 1.x and 2 on the same host.

File: server/src/main/java/com/paypal/selion/grid/AbstractBaseProcessLauncher.java [hunk 9]
```diff
@@ -200,7 +229,7 @@ void init(String[] args, ProcessLauncherOptions options) {
      * @throws InterruptedException
      */
     final void continuouslyRestart(long interval) throws IOException, InterruptedException {
-        LOGGER.entering(new Object[]{interval});
+        LOGGER.entering(new Object[] { interval });
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/AbstractBaseProcessLauncher.java]: minor but the array is not necessary here.

File: server/src/main/java/com/paypal/selion/grid/LauncherConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        noDownloadCleanup = !other.isFileDownloadCleanupOnInvocation();
+        noDownloadTimeStampCheck = !other.isFileDownloadCheckTimeStampOnInvocation();
+        if (StringUtils.isNotBlank(other.getSeLionConfig())) {
+            selionConfig = other.getSeLionConfig();
+        }
+    }
+
+    public void merge(LauncherConfiguration other) {
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/LauncherConfiguration.java]: Can't one method handle both this scenario and the one above?

File: server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java [hunk 11]
```diff
@@ -73,47 +74,31 @@ public MobileProcessLauncher(String[] args, ProcessLauncherOptions options) {
             }
         }
 
-        // filter out SeLion Grid specific args which do not apply
-        List<String> filteredArgs = new LinkedList<>();
-        filteredArgs.add(SELION_CONFIG_ARG);
-        filteredArgs.add(SELION_NOCONTINUOUS_ARG);
+        // filter out all SeLion Grid args -- They do not apply
+        // Assume all of the ProcessLauncherConfiguration and LauncherConfiguration arguments should not be forwarded
+        Set<Field> fields = new HashSet<Field>();
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java]: Can we use the diamond operator here for consistency.

File: server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java [hunk 12]
```diff
@@ -73,47 +74,31 @@ public MobileProcessLauncher(String[] args, ProcessLauncherOptions options) {
...
-                args.remove(args.indexOf(filter) + 1);
+        for (Field f : fields) {
+            Parameter p = f.getAnnotation(Parameter.class);
+            if (p == null) {
+                continue;
+            }
+            if (!f.getType().equals(Boolean.class)) {
+                for (String arg : args) {
+                    if (Arrays.asList(p.names()).contains(arg)) {
+                        args.set(args.indexOf(arg) + 1, "");
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java]: This block seems a little hard to follow what's happening here.

**Review by sebady — APPROVED** [review]
Have added mostly minor comments here.

File: server/src/main/java/com/paypal/selion/grid/JarSpawner.java (hunk 8 above)
Comment (mach6) [reply server/src/main/java/com/paypal/selion/grid/JarSpawner.java]: good suggestion!

File: server/pom.xml (hunk 7 above)
Comment (mach6) [reply server/pom.xml]: yep!

File: server/src/main/java/com/paypal/selion/grid/AbstractBaseProcessLauncher.java (hunk 9 above)
Comment (mach6) [reply server/src/main/java/com/paypal/selion/grid/AbstractBaseProcessLauncher.java]: fixed

File: server/src/main/java/com/paypal/selion/grid/LauncherConfiguration.java (hunk 10 above)
Comment (mach6) [reply server/src/main/java/com/paypal/selion/grid/LauncherConfiguration.java]: Hmm.. Yes. I figured this might come up.. To fix this I'll need to change the truthy nature of the `LauncherConfiguration` boolean `Parameters` which is not a problem.. Will fix.

File: server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java (hunk 11 above)
Comment (mach6) [reply server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java]: fixed

File: server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java (hunk 12 above)
Comment (mach6) [reply server/src/main/java/com/paypal/selion/grid/MobileProcessLauncher.java]: Will add some more comments

**Comment by mach6** [comment]
@sebady PR updated..
